### PR TITLE
Updates on workbench markers

### LINKED ISSUE #26724: Workbench Markers with empty label cause problems when editing them
### Expected behavior
The markers should always have a non empty and unique label to uniquely identify it when editing markers. A checkbox should allow the user to show/hide the label.

### Actual behavior
Labels can be duplicated or set to "".

### PR DESCRIPTION
**Pull request by BassoLorenzo**
**Description of work.**
 - The name of the marker can no longer be left empty
 - The name of the markers cannot be duplicated
 - An option is added to make the name visible/invisible
 - An option is added to make the marker draggable or not
 - Label positions can be modified

**To test:**
1. Open a plot
2. Add a marker, try to name it "", an error should appear
3. Create a second marker and give it the same name as the first. An error should appear
4. Check option `Fix position`, the marker cannot be dragged (but the position can be changed from the edit window)
5. Uncheck `Display label`, the marker label should disappear
6. Change the marker name and check `Display label`. The new label should appear.
7. Change the label position. The label should move.

Fixes #26724, #26725

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/applications/workbench/workbench/plotting/propertiesdialog.py [hunk 1]
```diff
@@ -199,25 +202,41 @@ def set_defaults(self, marker):
...
     def update_marker(self, marker):
         """
         Update the properties of the marker with the values from the widget
         """
         old_name = str(marker.name)
+        new_name = self.widget.name.text()
+        if new_name == "":
+            raise RuntimeError("Marker names cannot be empty")
+        if new_name in self.used_names and new_name != old_name:
+            raise RuntimeError("Marker names cannot be duplicated.\nAnother marker is named '{}'"
```
Comment (AnthonyLim23) [inline qt/applications/workbench/workbench/plotting/propertiesdialog.py]:
```suggestion
            raise RuntimeError("Marker names cannot be duplicated.\n Another marker is named '{}'"
```

File: qt/applications/workbench/workbench/plotting/propertiesdialog.py [hunk 2]
```diff
@@ -260,13 +279,17 @@ def __init__(self, canvas, markers, valid_style, valid_colors):
         self.ui.errors.hide()
         self.ui.marker.currentIndexChanged.connect(self.update_marker_data)
 
-        self._widget = MarkerEditor('markeredit.ui', valid_style, valid_colors)
+        self.markers = sorted(markers, key=lambda x: x.name)
```
Comment (AnthonyLim23) [inline qt/applications/workbench/workbench/plotting/propertiesdialog.py]: can we have a better name then x?

File: qt/python/mantidqt/plotting/markers.py [hunk 3]
```diff
@@ -786,25 +821,35 @@ def add_annotate(self, text):
...
+        if not self.label_visible:
+            self.annotations[self.name] = ''
+            return
+
         marker_in_scope = True
         x_lower, x_upper = self.marker.axis.get_xlim()
         y_lower, y_upper = self.marker.axis.get_ylim()
         if self.marker_type == 'YSingle':
+            x_pos = self.label_x_pos
+            y_pos = self.relative(self.marker.y, y_lower, y_upper) + self.label_y_pos
```
Comment (AnthonyLim23) [inline qt/python/mantidqt/plotting/markers.py]: it took me a while, but is `self.label_y_pos` actually the offset from the y position?

**Comment by AnthonyLim23** [comment]
open a tiled plot (with at least two subplots), add markers to the top left subplot (this is fine). When you use the context menu on the another subplot the lines are added to the first subplot

**Comment by AnthonyLim23** [comment]
just had a look and it now adds to the correct subplot. However, I can edit any marker from any subplot. Even if that marker is not on the current subplot. Is this an intended feature?

**Review by AnthonyLim23 — DISMISSED** [review]
:shipit:

File: qt/applications/workbench/workbench/plotting/propertiesdialog.py [hunk 4]
```diff
@@ -173,7 +173,7 @@ def __init__(self, canvas, axes):
 
 
 class MarkerEditor(QWidget):
-    def __init__(self, filename, valid_style, valid_colors):
+    def __init__(self, filename, valid_style, valid_colors, used_names=[]):
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plotting/propertiesdialog.py]: It is not best practice to use [mutable objects as defaults](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) in Python. Can we follow the standard guidance as suggested on the guide and use `None`?

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Comment by AnthonyLim23** [comment]
@martyngigg are you happy with this?

**Comment by martyngigg** [comment]
Sorry I missed that the update had happened. I'm happy now.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**